File: go/types/app/api_handler.go
package types_app

import "github.com/rs/zerolog"

type ApiHandler interface {
	CtxHandler
	Log() *zerolog.Logger
	Api() API

	ErrorMsg(err error)
}

type APIMsg[REQ any, RES any] interface {
	ApiHandler
	WriteMsg(data *RES)
	RequestMsg() (*REQ, error)
}

type ApiHandlerTyped[REQ any, RES any] func(api APIMsg[REQ, RES]) bool

type ApiHandlerUnTyped func(api ApiHandler) bool

File: ts/src/types/app/api_handler.ts
import { Logger } from "@adviser/cement";
import { Api } from "./api.js";
import { CtxHandler } from "./ctx_handler.js";

export interface ApiHandler extends CtxHandler {
  Log(): Logger; // *zerolog.Logger
  Api(): Api;

  ErrorMsg<T>(err: T | Error): Promise<number>;
}

export interface APIMsg<REQ, RES> extends ApiHandler {
  WriteMsg(data: RES): Promise<number>;
  RequestMsg(): Promise<REQ>;
}

export type ApiHandlerTyped<REQ, RES> = (api: APIMsg<REQ, RES>) => Promise<boolean>;

export type ApiHandlerUnTyped = (api: ApiHandler) => Promise<boolean>;
